cpp/src/arrow/compute/kernels/vector_hash.cc

```diff
@@ -76,6 +79,8 @@ class UniqueAction final : public ActionBase {
   template <class Index>
   void ObserveNotFound(Index index) {}
 
+  bool ShouldEncodeNulls() { return true; }
```

Comment:
Making this explicitly constexpr (where possible) will help the compiler drop dead code 
```suggestion
  constexpr bool ShouldEncodeNulls() const { return true; }
```